### ISSUE TITLE
Repo migration plan

### ISSUE THREAD

**Issue by mvines** [solana-labs/spl-zk-token#91]
The contents of this repo should be merged into the Solana and SPL mono repos.  Here's how:

#### Part I: https://github.com/solana-labs/solana additions:
* [X] `zk-token-sdk/`  (crate name: `solana-zk-token-sdk`) - contents of [/sdk](https://github.com/solana-labs/spl-zk-token/tree/master/sdk)
   * Dismissed alternatives: I don't think this sdk should be merged into the `solana-sdk` nor `solana-program`, and locating this crate under the SDK directory (eg`sdk/zk-token`) feels clunkly too
* [X] `programs/zk-token-proof/` (crate:name `solana-zk-token-proof-program`) - contents of [/proof-program](https://github.com/solana-labs/spl-zk-token/tree/master/proof-program)

#### Part II: https://github.com/solana-labs/solana-program-library additions:
* [X] The contents of [/program](https://github.com/solana-labs/spl-zk-token/tree/master/program) move into a SPL Token extension under `token/program/src/`
* [x] The contents of [/paper](https://github.com/solana-labs/spl-zk-token/tree/master/paper) move to `token/zk-token-paper/`
* [ ] The contents of [/demo](https://github.com/solana-labs/spl-zk-token/tree/master/demo) get incorporated into `token/cli`-support for zk tokens

**Comment by mvines** [comment]
Migration is based on https://github.com/solana-labs/spl-zk-token/releases/tag/migrate

**Comment by mvines** [comment]
ref: https://github.com/solana-labs/solana/pull/22305

**Comment by mvines** [comment]
"Part I" is now complete, although it won't be fully functional until https://github.com/solana-labs/solana/pull/21756 lands

**Comment by mvines** [comment]
Next steps here: migrate the unit tests over to spl-token

**Comment by CriesofCarrots** [comment]
@mvines , what's the best way to make changes like this (https://github.com/solana-labs/solana/pull/22532) available in SPL? Should it be duplicated in this repo and the crate published with a bumped version?

**Comment by mvines** [comment]
As a temporarily measure until we either bring the crate into v1.9 or v1.10 starts shipping I pushed v0.1.0 to https://crates.io/crates/solana-zk-token-sdk so to manually edit the package.json on the monorepo master and then `cargo publish` locally.  I can do this tomorrow morning again if you'd like?

**Comment by CriesofCarrots** [comment]
> I can do this tomorrow morning again if you'd like?

Thank you, that would be great!

**Comment by mvines** [comment]
I published v0.2.0 of solana-zk-token-sdk based on https://github.com/solana-labs/solana/commit/e3915e4b7ac7092f8864a356266addea2762328f with this patch
```diff
diff --git a/zk-token-sdk/Cargo.toml b/zk-token-sdk/Cargo.toml
index eac149b617..bc41af5eb9 100644
--- a/zk-token-sdk/Cargo.toml
+++ b/zk-token-sdk/Cargo.toml
@@ -3,7 +3,7 @@ name = "solana-zk-token-sdk"
 description = "Solana Zk Token SDK"
 authors = ["Solana Maintainers <maintainers@solana.foundation>"]
 repository = "https://github.com/solana-labs/solana"
-version = "1.10.0"
+version = "0.2.0"
 license = "Apache-2.0"
 edition = "2021"
 
@@ -12,7 +12,7 @@ base64 = "0.13"
 bytemuck = { version = "1.7.2", features = ["derive"] }
 num-derive = "0.3"
 num-traits = "0.2"
-solana-program = { path = "../sdk/program", version = "=1.10.0" }
+solana-program = { path = "../sdk/program", version = "=1.9.2" }
 
 [target.'cfg(not(target_arch = "bpf"))'.dependencies]
 aes-gcm-siv = "0.10.3"
@@ -28,7 +28,7 @@ rand = "0.7"
 serde = { version = "1.0", features = ["derive"] }
 serde_json = "1.0"
 sha3 = "0.9"
-solana-sdk = { path = "../sdk", version = "=1.10.0" }
+solana-sdk = { path = "../sdk", version = "=1.9.2" }
 subtle = "2"
 thiserror = "1"
 zeroize = { version = "1.2.0", default-features = false, features = ["zeroize_derive"] }
```

**Comment by CriesofCarrots** [comment]
Okay if I publish a v0.3.0 with https://github.com/solana-labs/solana/pull/22639 ?
(Can wait on https://github.com/solana-labs/solana/pull/22632 if desired)

**Comment by mvines** [comment]
I see no reason to wait, there's plenty of version numbers.

**Comment by CriesofCarrots** [comment]
Note to future readers: the patch above (https://github.com/solana-labs/spl-zk-token/issues/91#issuecomment-1018693188) needs to specify solana-program and solana-sdk versions as caret requirements instead of exact (eg. `1.9.2`, not `=1.9.2`). Otherwise, it will break the monorepo downstream-projects CI build when SPL is bumped

**Comment by mvines** [comment]
Mostly migrated to https://github.com/solana-labs/solana-program-library/tree/master/token/program-2022/src/extension/confidential_transfer now.   All that's left is bringing over some of the unit tests and `demo/` program

**Comment by mvines** [comment]
Basically done now
